### PR TITLE
V2 DXF Serializer

### PR DESCRIPTION
**Pull request by z3dev**
This pull request updates the DXF serializer to use CSG V2 API. The goals of this pull request are:
- [X] Update to use CSG V2 API
- [X] Update code up to JSCAD coding standards
- [X] Add test suites to complete coverage, 97.08%

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

### CODE REVIEW

File: packages/io/dxf-serializer/.gitignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+node_modules
```
Comment (kaosat-dev) [inline packages/io/dxf-serializer/.gitignore]: gitignore not needed :)

File: packages/io/dxf-serializer/README.md [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
-> DXF serializer for the JSCAD project (from CSG)
+> Serializer of CSG geometries to DXF entities
 
 [![npm version](https://badge.fury.io/js/%40jscad%2Fdxf-serializer.svg)](https://badge.fury.io/js/%40jscad%2Fdxf-serializer)
 [![Build Status](https://travis-ci.org/jscad/io.svg)](https://travis-ci.org/jscad/dxf-serializer)
 
 ## Overview
 
-This serializer outputs a 'blobable' array of data (from one or more CSG library objects).
+This serializer outputs a 'blobable' array of data from one or more CSG geometries. Currently, only TEXT output is supported.
```
Comment (kaosat-dev) [inline packages/io/dxf-serializer/README.md]: oh, I did not know that only text is supported, good to know (I understand given the complexity!)
Btw perhaps it should perhaps be JSCAD script TEXT output ?

File: packages/io/dxf-serializer/index.js [hunk 3]
```diff
@@ -8,56 +8,48 @@ Copyright (c) 2018 Z3 Development https://github.com/z3dev
...
-  }
-  // make sure we always deal with arrays of objects as inputs
-  objects = toArray(objects)
-
+const serialize = (options, ...objects) => {
   const defaults = {
     cagTo: 'lwpolyline', // or polyline
     csgTo: '3dface', // or polyline
     pathTo: 'lwpolyline',
+    entityId: 0,
```
Comment (kaosat-dev) [inline packages/io/dxf-serializer/index.js]: what does this one do ?

File: packages/io/dxf-serializer/index.js [hunk 4]
```diff
@@ -8,56 +8,48 @@ Copyright (c) 2018 Z3 Development https://github.com/z3dev
...
     statusCallback: null
   }
   options = Object.assign({}, defaults, options)
 
+  if (objects.length === 0) {
+    throw new Error('no arguments supplied to serialize function !')
```
Comment (kaosat-dev) [inline packages/io/dxf-serializer/index.js]: not arguments, objects

File: packages/io/dxf-serializer/index.js [hunk 5]
```diff
@@ -78,27 +70,25 @@ EOF
  * @returns {Object} serialized contents, DXF format
  */
 const dxfEntities = (objects, options) => {
-  objects = toArray(objects)
-  let entityContents = objects.map(function (object, i) {
-    if (isCAG(object)) {
-      let paths = object.getOutlinePaths()
+  let entityContents = objects.map((object, i) => {
+    if (geometry.geom2.isA(object)) {
```
Comment (kaosat-dev) [inline packages/io/dxf-serializer/index.js]: should this not be filtered like you did in a few other tools ?

File: packages/io/dxf-serializer/index.js [hunk 6]
```diff
@@ -78,27 +70,25 @@ EOF
...
-    // mimic a CAG outline path
-      let points = object.getPoints()
-      let closed = object.isClosed()
-      let path = {closed: closed, points: points}
+    if (geometry.path2.isA(object)) {
+      // mimic a path (outline) from geom2
+      let path = { closed: object.isClosed, points: geometry.path2.toPoints(object) }
       return PathsToLwpolyine([path], options)
     }
     return ''
```
Comment (kaosat-dev) [inline packages/io/dxf-serializer/index.js]: ah ok, I guess it makes sense t

File: packages/io/dxf-serializer/index.js [hunk 7]
```diff
@@ -119,7 +109,7 @@ ENDSEC`
 }
 
 //
-// convert the given paths (from CAG outlines) to DXF lwpolyline entities
+// convert the given paths (from 2D outlines) to DXF lwpolyline entities
```
Comment (kaosat-dev) [inline packages/io/dxf-serializer/index.js]: good

File: packages/io/dxf-serializer/index.js [hunk 8]
```diff
@@ -259,14 +249,14 @@ const polygonToTriangles = (polygon) => {
 // convert the given triangle to DXF 3D face entity
 //
 const triangleTo3DFaces = (triangle, options) => {
-  let corner10 = triangle[0].pos
-  let corner11 = triangle[1].pos
-  let corner12 = triangle[2].pos
-  let corner13 = triangle[2].pos // same in DXF
+  let corner10 = triangle[0]
```
Comment (kaosat-dev) [inline packages/io/dxf-serializer/index.js]: haha so much simpler

File: packages/io/dxf-serializer/.gitignore (hunk 1 above)
Comment (z3dev) [reply packages/io/dxf-serializer/.gitignore]: Done

File: packages/io/dxf-serializer/README.md (hunk 2 above)
Comment (z3dev) [reply packages/io/dxf-serializer/README.md]: Done

File: packages/io/dxf-serializer/index.js (hunk 3 above)
Comment (z3dev) [reply packages/io/dxf-serializer/index.js]: Removed from defaults, as the entityId is used to control processing.

File: packages/io/dxf-serializer/index.js (hunk 4 above)
Comment (z3dev) [reply packages/io/dxf-serializer/index.js]: Done

File: packages/io/dxf-serializer/index.js (hunk 5 above)
Comment (z3dev) [reply packages/io/dxf-serializer/index.js]: Added filtering. And error.

File: packages/io/dxf-serializer/index.js [hunk 9]
```diff
@@ -8,48 +8,34 @@ Copyright (c) 2018 Z3 Development https://github.com/z3dev
...
-  } else if (params.length > 1) {
-    options = params[0]
-    objects = params[1]
-  }
-  // make sure we always deal with arrays of objects as inputs
-  objects = toArray(objects)
-
+const serialize = (options, ...objects) => {
   const defaults = {
     cagTo: 'lwpolyline', // or polyline
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/io/dxf-serializer/index.js]: Done

File: packages/io/dxf-serializer/index.js [hunk 10]
```diff
@@ -39,11 +40,12 @@ const serialize = (options, ...objects) => {
...
     csgTo: '3dface', // or polyline
     pathTo: 'lwpolyline',
-    entityId: 0,
     statusCallback: null
   }
   options = Object.assign({}, defaults, options)
 
+  options.entityId = 0 // sequence id for entities created
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/io/dxf-serializer/index.js]: Yes. This is a lazy man's control variable for sequential entity ids.

File: packages/io/dxf-serializer/tests/path2ToDxf.test.js [hunk 11]
```diff
@@ -1,32 +1,33 @@
...
 
   const obs2 = serialize({}, path2)
   const exp2 = [lwpolyline0]
   t.deepEqual(obs2, exp2)
 
-  let path3 = new CSG.Path2D([[10, -20]], false)
-  path3 = path3.appendBezier([[10, -10], [25, -10], [25, -20]], { resolution: 8 })
-  t.is(path3.points.length, 6)
+  let path3 = geometry.path2.fromPoints({}, [[10, -20]])
+// TODO
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/io/dxf-serializer/tests/path2ToDxf.test.js]: Nope. There's a test for bezier curves that's incomplete. Not really required but a nice to do.

**Comment by z3dev** [comment]
@kaosat-dev please review, and merge if no futher changes are required.

File: packages/io/dxf-serializer/tests/path2ToDxf.test.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (kaosat-dev) [reply packages/io/dxf-serializer/tests/path2ToDxf.test.js]: ok

**Pull request merged**